### PR TITLE
Tensor compression (part1)

### PR DESCRIPTION
**Pull request by Vsevolod-pl**

### CODE REVIEW

File: hivemind/server/connection_handler.py [hunk 1]
```diff
@@ -58,13 +58,18 @@ async def info(self, request: runtime_pb2.ExpertUID, context: grpc.ServicerConte
         return runtime_pb2.ExpertInfo(serialized_info=pickle.dumps(self.experts[request.uid].get_info()))
 
     async def forward(self, request: runtime_pb2.ExpertRequest, context: grpc.ServicerContext):
+        # TODO: add loading compression parameter
         inputs = [deserialize_torch_tensor(tensor) for tensor in request.tensors]
         future = self.experts[request.uid].forward_pool.submit_task(*inputs)
-        serialized_response = [serialize_torch_tensor(tensor) for tensor in await future]
+        forward_schema_flatten = list(nested_flatten(self.experts[request.uid].forward_schema))
+        serialized_response = [serialize_torch_tensor(tensor, forward_schema_flatten[i].compression)
```
Comment (justheuristic) [inline hivemind/server/connection_handler.py:65]: Q: can we use iterator over zip here (instead of creating a list)
This should theoretically be faster and arguably more elegant

File: hivemind/server/connection_handler.py [hunk 2]
```diff
@@ -58,13 +58,18 @@ async def info(self, request: runtime_pb2.ExpertUID, context: grpc.ServicerConte
...
+                               for i, tensor in enumerate(await future)]
         return runtime_pb2.ExpertResponse(tensors=serialized_response)
 
     async def backward(self, request: runtime_pb2.ExpertRequest, context: grpc.ServicerContext):
         inputs_and_grad_outputs = [deserialize_torch_tensor(tensor) for tensor in request.tensors]
         future = self.experts[request.uid].backward_pool.submit_task(*inputs_and_grad_outputs)
-        serialized_response = [serialize_torch_tensor(tensor) for tensor in await future]
+        backward_schema_flatten = list(nested_flatten(self.experts[request.uid].backward_schema))
+        serialized_response = [serialize_torch_tensor(tensor, backward_schema_flatten[i].compression)
+                               for i, tensor in enumerate(await future)]
```
Comment (justheuristic) [inline hivemind/server/connection_handler.py:74]: (and here as well)

File: hivemind/utils/grpc.py [hunk 3]
```diff
@@ -6,19 +6,44 @@
...
-        buffer=array.tobytes(),
-        size=array.shape,
-        dtype=array.dtype.name,
-        requires_grad=tensor.requires_grad)
+    if compression_type == CompressionType.MEANSTD_LAST_AXIS_FLOAT16:
+        assert array.dtype == np.float32
+
+        mean = array.mean()
+        std = array.std()
+        normalized = (array - mean) / std
```
Comment (justheuristic) [inline hivemind/utils/grpc.py:19]: Request: you're using numpy mean/std, but torch also has its own mean/std functions. Can you please test which one is faster (on cpu)?

File: hivemind/utils/grpc.py [hunk 4]
```diff
@@ -6,19 +6,44 @@
...
-    proto = runtime_pb2.Tensor(
-        buffer=array.tobytes(),
-        size=array.shape,
-        dtype=array.dtype.name,
-        requires_grad=tensor.requires_grad)
+    if compression_type == CompressionType.MEANSTD_LAST_AXIS_FLOAT16:
+        assert array.dtype == np.float32
+
+        mean = array.mean()
+        std = array.std()
```
Comment (justheuristic) [inline hivemind/utils/grpc.py:18]: i'm unsure but this may need axis=-1, keepdim=True.

File: hivemind/utils/grpc.py (hunk 3 above)
Comment (justheuristic) [reply hivemind/utils/grpc.py:19]: Third option: make it a three-liner:
- compute mean
- subtract mean in-place
- compute std = euclidean norm, since we already subtracted mean
- divide by std

**Review by justheuristic — CHANGES_REQUESTED** [review]

**Comment by Maria1007** [comment]
Hi！
What should I do？

File: .circleci/config.yml [hunk 5]
```diff
@@ -32,4 +32,4 @@ jobs:
 workflows:
   main:
     jobs:
-      - build-and-test
+      - build-and-test
```
Comment (justheuristic) [inline .circleci/config.yml:35]:
```suggestion
      - build-and-test

```

File: hivemind/server/connection_handler.py [hunk 6]
```diff
@@ -58,13 +58,18 @@ async def info(self, request: runtime_pb2.ExpertUID, context: grpc.ServicerConte
         return runtime_pb2.ExpertInfo(serialized_info=pickle.dumps(self.experts[request.uid].get_info()))
 
     async def forward(self, request: runtime_pb2.ExpertRequest, context: grpc.ServicerContext):
+        # TODO: add loading compression parameter
         inputs = [deserialize_torch_tensor(tensor) for tensor in request.tensors]
         future = self.experts[request.uid].forward_pool.submit_task(*inputs)
-        serialized_response = [serialize_torch_tensor(tensor) for tensor in await future]
+        forward_schema_flatten = list(nested_flatten(self.experts[request.uid].forward_schema))
+        serialized_response = [serialize_torch_tensor(tensor, forward_schema_flatten[i].compression)
+                               for i, tensor in enumerate(await future)]
```
Comment (justheuristic) [inline hivemind/server/connection_handler.py:66]:
```suggestion
        forward_schema_flatten = list(nested_flatten(self.experts[request.uid].forward_schema))
        serialized_response = [serialize_torch_tensor(tensor, proto.compression) tensor, proto in zip(
                               await future, nested_flatten(self.experts[request.uid].forward_schema))]
```

Can we re-write this with iterators instead of instantiating a list?

File: hivemind/utils/grpc.py [hunk 7]
```diff
@@ -6,19 +6,53 @@
...
+
+        tensor = tensor if allow_inplace else tensor.clone()
+        means = torch.mean(tensor, dim=-1, keepdim=True)
+        tensor.sub_(means)
+
+        stds = torch.square(tensor).sum(dim=-1, keepdim=True).div_(tensor.shape[-1]).sqrt_()
+        tensor.div_(stds)
+        tensor = tensor.clamp_(-FP16_MAX, FP16_MAX).to(torch.float16)
+
+        data = tensor.numpy().tobytes() + means.numpy().tobytes() + stds.numpy().tobytes()
```
Comment (justheuristic) [inline hivemind/utils/grpc.py:25]:
```suggestion
        data = b''.join((tensor.numpy().tobytes(), means.numpy().tobytes(), stds.numpy().tobytes()))
```

Comment (justheuristic) [reply hivemind/utils/grpc.py:25]: what's the difference:

* x+y+z allocates new memory after each addition (i.e. twice)
* b''.join allocates memory once

File: hivemind/utils/grpc.py [hunk 8]
```diff
@@ -6,19 +6,53 @@
...
+
+        proto = runtime_pb2.Tensor(
+            compression=compression_type,
+            buffer=data,
+            size=tensor.shape,
+            dtype='compressed_float16',
```
Comment (justheuristic) [inline hivemind/utils/grpc.py:31]: (minor) maybe float32? reason: the tensor contains float32 values, but it is compressed to other dtype, i.e. float32 is compressed, not float16.

File: hivemind/utils/grpc.py [hunk 9]
```diff
@@ -6,19 +6,57 @@
...
+        stats_size = compressed_tensor.size
+        stats_size[-1] = 1
+        stats_count = 1
+        for i in stats_size:
+            stats_count *= i
+        means, stds = compressed_tensor.buffer[-8*stats_count:-4*stats_count], compressed_tensor.buffer[-4*stats_count:]
+        means = torch.as_tensor(np.frombuffer(means, dtype=np.float32)).reshape(stats_size)
+        stds = torch.as_tensor(np.frombuffer(stds, dtype=np.float32)).reshape(stats_size)
+        array = np.frombuffer(compressed_tensor.buffer[:-8 * stats_count], dtype=np.float16)
+        tensor = torch.as_tensor(array).to(torch.float32).view(*compressed_tensor.size).mul_(stds).add_(means)
```
Comment (justheuristic) [inline hivemind/utils/grpc.py:61]: пожелание от pycharm:
else:
    raise ValueError(f"Unknown compression type: {compressed_tensor.compression}")

File: hivemind/utils/grpc.py [hunk 10]
```diff
@@ -6,19 +6,57 @@
...
+
     return proto
 
 
-def deserialize_torch_tensor(tensor: runtime_pb2.Tensor) -> torch.Tensor:
-    # TODO avoid copying the array (need to silence pytorch warning, because array is not writable)
-    array = np.frombuffer(tensor.buffer, dtype=np.dtype(tensor.dtype)).copy()
-    return torch.as_tensor(array).view(tuple(tensor.size)).requires_grad_(tensor.requires_grad)
+def deserialize_torch_tensor(compressed_tensor: runtime_pb2.Tensor) -> torch.Tensor:
+    # TODO avoid copying the array (need to silence pytorch warning,x because array is not writable)
```
Comment (justheuristic) [inline hivemind/utils/grpc.py:46]:
```suggestion
    # TODO avoid copying the array (need to silence pytorch warning, because array is not writable)
```

File: hivemind/utils/grpc.py [hunk 11]
```diff
@@ -6,19 +6,57 @@
...
+    if compressed_tensor.compression == CompressionType.NONE:
+        array = np.frombuffer(compressed_tensor.buffer, dtype=np.dtype(compressed_tensor.dtype)).copy()
+        array = array.reshape(tuple(compressed_tensor.size))
+        tensor = torch.as_tensor(array).requires_grad_(compressed_tensor.requires_grad)
```
Comment (justheuristic) [inline hivemind/utils/grpc.py:50]:
```suggestion
        tensor = torch.as_tensor(array)..view(*compressed_tensor.size).requires_grad_(compressed_tensor.requires_grad)
```

File: hivemind/utils/grpc.py [hunk 12]
```diff
@@ -6,19 +6,57 @@
...
+        stats_size[-1] = 1
+        stats_count = 1
+        for i in stats_size:
+            stats_count *= i
+        means, stds = compressed_tensor.buffer[-8*stats_count:-4*stats_count], compressed_tensor.buffer[-4*stats_count:]
+        means = torch.as_tensor(np.frombuffer(means, dtype=np.float32)).view(*stats_size)
+        stds = torch.as_tensor(np.frombuffer(stds, dtype=np.float32)).view(*stats_size)
+        array = np.frombuffer(compressed_tensor.buffer[:-8 * stats_count], dtype=np.float16)
+        tensor = torch.as_tensor(array).to(torch.float32).view(*compressed_tensor.size).mul_(stds).add_(means)
+    return tensor
```
Comment (justheuristic) [inline hivemind/utils/grpc.py:62]:
```suggestion
    else:
        raise ValueError(f"Unknown compression type: {compressed_tensor.compression}")
    return tensor
```

rationale: if you received a compression type that you do not understand, its better to raise explicit error than "Undefined variable tensor"

File: hivemind/utils/grpc.py [hunk 13]
```diff
@@ -6,19 +6,55 @@
...
-def deserialize_torch_tensor(tensor: runtime_pb2.Tensor) -> torch.Tensor:
-    # TODO avoid copying the array (need to silence pytorch warning, because array is not writable)
-    array = np.frombuffer(tensor.buffer, dtype=np.dtype(tensor.dtype)).copy()
-    return torch.as_tensor(array).view(tuple(tensor.size)).requires_grad_(tensor.requires_grad)
+def deserialize_torch_tensor(serialized_tensor: runtime_pb2.Tensor) -> torch.Tensor:
+    # TODO avoid copying the array (need to silence pytorch warning,x because array is not writable)
+    if serialized_tensor.compression == CompressionType.NONE:
+        array = np.frombuffer(serialized_tensor.buffer, dtype=np.dtype(serialized_tensor.dtype)).copy()
+        array = array.reshape(tuple(serialized_tensor.size))
+        tensor = torch.as_tensor(array).requires_grad_(serialized_tensor.requires_grad)
```
Comment (justheuristic) [inline hivemind/utils/grpc.py:50]:
```suggestion
        tensor = torch.as_tensor(array).view(*compressed_tensor.size).requires_grad_(compressed_tensor.requires_grad)
```

**Review by justheuristic — APPROVED** [review]

File: hivemind/utils/grpc.py [hunk 14]
```diff
@@ -6,19 +6,55 @@
...
 
 
-def serialize_torch_tensor(tensor: torch.Tensor) -> runtime_pb2.Tensor:
-    array = tensor.numpy()
-    proto = runtime_pb2.Tensor(
-        buffer=array.tobytes(),
-        size=array.shape,
-        dtype=array.dtype.name,
-        requires_grad=tensor.requires_grad)
+def serialize_torch_tensor(tensor: torch.Tensor, compression_type=CompressionType.NONE, allow_inplace=False) -> runtime_pb2.Tensor:
```
Comment (justheuristic) [inline hivemind/utils/grpc.py:12]:
```suggestion
def serialize_torch_tensor(tensor: torch.Tensor, compression_type=CompressionType.NONE, 
                           allow_inplace=False) -> runtime_pb2.Tensor:
```

File: hivemind/server/connection_handler.py [hunk 15]
```diff
@@ -60,11 +60,13 @@ async def info(self, request: runtime_pb2.ExpertUID, context: grpc.ServicerConte
     async def forward(self, request: runtime_pb2.ExpertRequest, context: grpc.ServicerContext):
         inputs = [deserialize_torch_tensor(tensor) for tensor in request.tensors]
         future = self.experts[request.uid].forward_pool.submit_task(*inputs)
-        serialized_response = [serialize_torch_tensor(tensor) for tensor in await future]
+        serialized_response = [serialize_torch_tensor(tensor, proto.compression) for tensor, proto in zip(
+            await future, nested_flatten(self.experts[request.uid].outputs_schema))]
```
Comment (justheuristic) [inline hivemind/server/connection_handler.py:64]:
```suggestion
        serialized_response = [serialize_torch_tensor(tensor, proto.compression, allow_inplace=True) for tensor, proto
                               in zip(await future, nested_flatten(self.experts[request.uid].outputs_schema))]

```

rationale: added allow_inplace=True

File: hivemind/server/connection_handler.py [hunk 16]
```diff
@@ -60,11 +60,13 @@ async def info(self, request: runtime_pb2.ExpertUID, context: grpc.ServicerConte
...
         return runtime_pb2.ExpertResponse(tensors=serialized_response)
 
     async def backward(self, request: runtime_pb2.ExpertRequest, context: grpc.ServicerContext):
         inputs_and_grad_outputs = [deserialize_torch_tensor(tensor) for tensor in request.tensors]
         future = self.experts[request.uid].backward_pool.submit_task(*inputs_and_grad_outputs)
-        serialized_response = [serialize_torch_tensor(tensor) for tensor in await future]
+        serialized_response = [serialize_torch_tensor(tensor, proto.compression) for tensor, proto in zip(
+            await future, nested_flatten(self.experts[request.uid].grad_inputs_schema))]
```
Comment (justheuristic) [inline hivemind/server/connection_handler.py:71]:
```suggestion
        serialized_response = [serialize_torch_tensor(tensor, proto.compression, allow_inplace=True) for tensor, proto
                               in zip(await future, nested_flatten(self.experts[request.uid].grad_inputs_schema))]
```

(added allow_inplace=True)

File: hivemind/utils/grpc.py (hunk 8 above)
Comment (justheuristic) [reply hivemind/utils/grpc.py:31]: image
